org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/testdetection/TestDetectionHandler.java

```diff
@@ -22,20 +28,38 @@
 import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.ls.core.internal.JDTUtils;
 import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
 
 /** Handler for test detection events. */
 public class TestDetectionHandler {
 
+  private static final Gson gson =
+      new GsonBuilder()
+          .registerTypeAdapterFactory(new CollectionTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EitherTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
+          .create();
+
   /**
    * Detects if the java class has tests.
    *
    * @param arguments a list contains file URI, fqn of test method annotation and cursor offset
+   * @param pm a progress monitor
    * @return test positions @see {@link TestPosition}
    */
-  public static List<TestPosition> detect(List<Object> arguments) {
-    String fileUri = (String) arguments.get(0);
-    String testAnnotation = (String) arguments.get(1);
-    int cursorOffset = ((Double) arguments.get(2)).intValue();
+  public static List<TestPosition> detect(List<Object> arguments, IProgressMonitor pm) {
+    TestPositionParameters parameters =
+        gson.fromJson(gson.toJson(arguments.get(0)), TestPositionParameters.class);
+
+    String fileUri = parameters.getFileUri();
+    String testAnnotation = parameters.getTestAnnotation();
+    int cursorOffset = parameters.getCursorOffset();
+
+    if (pm.isCanceled()) {
```

Comment:
Again, check cancellation where the real work is  done. It sucks for the user to press cancel and nothing happens.

Reply:
ok, done